File: AIV-Backend-mvp-vault/_archive/schemas/clone.py
from pydantic import BaseModel, Field
from typing import Optional, Dict, List, Any
from uuid import UUID
from datetime import datetime
from enum import Enum


class CloneStatusEnum(str, Enum):
    DRAFT = "draft"
    PROCESSING = "processing"
    COMPLETED = "completed"
    FAILED = "failed"


# ============== Request Schemas ==============

class CloneCreate(BaseModel):
    """Schema for creating a new clone."""
    name: str = Field(..., min_length=2, max_length=255)
    description: Optional[str] = None


class CloneUpdate(BaseModel):
    """Schema for updating clone metadata."""
    name: Optional[str] = Field(None, min_length=2, max_length=255)
    description: Optional[str] = None
    avatar_url: Optional[str] = None


class VoiceDataUpdate(BaseModel):
    """Schema for updating voice data (Stage 1)."""
    sample_url: str
    duration: Optional[int] = None
    transcription: Optional[str] = None


class NarrativeDataUpdate(BaseModel):
    """Schema for updating narrative answers (Stage 2).
    
    Contains answers to the personality questions.
    """
    answers: Dict[str, str] = Field(..., description="Dictionary of question_id: answer pairs")


class KnowledgeFileData(BaseModel):
    """Schema for a knowledge file."""
    filename: str
    url: str
    file_type: str
    size_bytes: Optional[int] = None


class KnowledgeDataUpdate(BaseModel):
    """Schema for updating knowledge files (Stage 3)."""
    files: List[KnowledgeFileData]


class ImageDataUpdate(BaseModel):
    """Schema for updating image data (Stage 4)."""
    frontal: Optional[str] = Field(None, description="URL to frontal portrait")
    profile: Optional[str] = Field(None, description="URL to profile angle photo")
    body: Optional[str] = Field(None, description="URL to full body photo")


class RightsUpdate(BaseModel):
    """Schema for updating privacy rights (Stage 5)."""
    is_public: bool = False
    allow_training: bool = False


class RawInputUpdate(BaseModel):
    """Schema for simplified single-paragraph onboarding."""
    raw_input: str = Field(..., min_length=50, description="Single paragraph describing yourself")


class DimensionData(BaseModel):
    """Schema for a single dimension."""
    content: str
    source: str = "initial"  # or "training"
    updated_at: str


class DimensionsResponse(BaseModel):
    """Response for clone dimensions."""
    version: int
    last_updated: Optional[datetime] = None
    dimensions: Dict[str, DimensionData]


class DimensionUpdate(BaseModel):
    """Schema for updating a single dimension."""
    content: str = Field(..., min_length=10)


# ============== Response Schemas ==============

class CloneResponse(BaseModel):
    """Full clone data response."""
    id: UUID
    owner_id: Optional[UUID]
    name: str
    description: Optional[str]
    avatar_url: Optional[str]
    status: CloneStatusEnum
    is_public: bool
    
    # Portal data
    voice_data: Optional[Dict[str, Any]]
    image_data: Optional[Dict[str, Any]]
    narrative_data: Optional[Dict[str, Any]]
    knowledge_files: Optional[List[Dict[str, Any]]]
    
    # ElevenLabs voice data
    elevenlabs_voice_id: Optional[str] = None
    intro_audio_url: Optional[str] = None
    
    # AI-generated avatars
    avatar_profile_url: Optional[str] = None
    avatar_icon_url: Optional[str] = None
    
    # Synthesized data (after activation)
    system_prompt: Optional[str]
    personality: Optional[Dict[str, Any]]
    background: Optional[str]
    
    # Simplified onboarding dimensions
    raw_input: Optional[str] = None
    dimensions: Optional[Dict[str, Any]] = None
    dimensions_version: Optional[int] = None
    
    # 3-Step Onboarding
    onboard_video_url: Optional[str] = None
    onboard_transcription: Optional[str] = None
    onboard_completed_at: Optional[datetime] = None
    
    created_at: datetime
    updated_at: datetime
    
    class Config:
        from_attributes = True


class CloneStatusResponse(BaseModel):
    """Response for checking portal completion status."""
    has_clone: bool
    clone_id: Optional[UUID]
    status: Optional[CloneStatusEnum]
    is_portal_complete: bool
    
    # Stage completion flags
    voice_complete: bool = False
    personality_complete: bool = False
    knowledge_complete: bool = False
    visual_complete: bool = False
    rights_complete: bool = False
    activated: bool = False

File: AIV-frontend-mvp-vault/_archive/lib/api/clone.ts
import { apiClient } from "./client";

export interface CloneStatus {
    has_clone: boolean;
    clone_id: string | null;
    status: string | null;
    is_portal_complete: boolean;
    voice_complete?: boolean;
    personality_complete?: boolean;
    knowledge_complete?: boolean;
    visual_complete?: boolean;
    rights_complete?: boolean;
    activated?: boolean;
    // New onboard status
    onboard_status?: 'pending' | 'processing' | 'complete' | 'failed';
}

// NEW: Onboard processing status interface
export interface OnboardStatus {
    overall_status: 'pending' | 'processing' | 'complete' | 'failed';
    video_saved: boolean;
    frames_extracted: boolean;
    audio_extracted: boolean;
    transcription_complete: boolean;
    dimensions_extracted: boolean;
    voice_cloned: boolean;
    avatar_generated: boolean;
    personality_synthesized: boolean;
    // Available when processing completes
    transcription?: string;
    dimensions?: Record<string, unknown>;
    voice_id?: string;
    avatar_url?: string;
    error?: string;
}

export interface CloneData {
    id: string;
    name: string;
    description?: string;
    status: string;
    voice_data?: {
        sampleUrl: string;
        duration: number;
    };
    narrative_data?: Record<string, string>;
    knowledge_files?: Array<{ name: string; url: string; size: number }>;
    image_data?: {
        frontal: string | null;
        profile: string | null;
        body: string | null;
    };
    is_public?: boolean;
    elevenlabs_voice_id?: string;
    intro_audio_url?: string;
    avatar_profile_url?: string;
    avatar_icon_url?: string;
    // AI-synthesized personality data
    system_prompt?: string;
    personality?: {
        traits?: string[];
        speaking_style?: string;
        quirks?: string[];
        knowledge_areas?: string[];
        values?: string[];
    };
    background?: string;
    // Onboard transcription from video
    onboard_transcription?: string;
    // 10 dimensions from onboard
    dimensions?: Record<string, unknown>;
    created_at: string;
    updated_at: string;
}

export interface CreateCloneData {
    name: string;
    description?: string;
}

export const cloneApi = {
    // Get clone status (for routing)
    getStatus: async (): Promise<CloneStatus> => {
        const response = await apiClient.get("/clone/status");
        return response.data;
    },

    // Create a new draft clone
    create: async (data: CreateCloneData): Promise<CloneData> => {
        const response = await apiClient.post("/clone", data);
        return response.data;
    },

    // Get clone details
    getClone: async (cloneId: string): Promise<CloneData> => {
        const response = await apiClient.get(`/clone/${cloneId}`);
        return response.data;
    },

    // ============================================
    // NEW: Video-based Onboard APIs
    // ============================================

    // Submit video onboard (multipart/form-data)
    submitOnboard: async (cloneId: string, formData: FormData): Promise<{ status: string; clone_id: string }> => {
        const response = await apiClient.post(`/clone/${cloneId}/onboard`, formData, {
            headers: {
                'Content-Type': 'multipart/form-data',
            },
        });
        return response.data;
    },

    // Get onboard processing status (poll every 2 seconds)
    getOnboardStatus: async (cloneId: string): Promise<OnboardStatus> => {
        const response = await apiClient.get(`/clone/${cloneId}/onboard/status`);
        return response.data;
    },

    // ============================================
    // Enhancement APIs (Post-Onboarding)
    // ============================================

    // Upload photos to enhance avatar (max 3 photos)
    enhancePhotos: async (cloneId: string, photos: File[]): Promise<{
        status: string;
        avatar_url?: string;
        message?: string;
    }> => {
        const formData = new FormData();
        photos.forEach((photo) => {
            formData.append('photos', photo);
        });
        const response = await apiClient.post(`/clone/${cloneId}/enhance/photos`, formData, {
            headers: {
                'Content-Type': 'multipart/form-data',
            },
        });
        return response.data;
    },

    // Upload knowledge documents to enhance clone
    enhanceKnowledge: async (cloneId: string, files: File[]): Promise<{
        status: string;
        files_uploaded: number;
        total_files: number;
    }> => {
        const formData = new FormData();
        files.forEach((file) => {
            formData.append('files', file);
        });
        const response = await apiClient.post(`/clone/${cloneId}/enhance/knowledge`, formData, {
            headers: {
                'Content-Type': 'multipart/form-data',
            },
        });
        return response.data;
    },

    // ============================================
    // Legacy APIs (kept for backward compatibility)
    // ============================================

    // Update voice data (Step 1)
    updateVoice: async (cloneId: string, data: { sample_url: string; duration: number }): Promise<void> => {
        await apiClient.put(`/clone/${cloneId}/voice`, data);
    },

    // Update personality/narrative data (Step 2)
    updatePersonality: async (cloneId: string, data: Record<string, string>): Promise<void> => {
        await apiClient.put(`/clone/${cloneId}/personality`, { answers: data });
    },

    // Update knowledge files (Step 3)
    updateKnowledge: async (cloneId: string, files: Array<{ name: string; url: string; size: number }>): Promise<void> => {
        await apiClient.put(`/clone/${cloneId}/knowledge`, { files });
    },

    // Update visual/images (Step 4)
    updateVisual: async (cloneId: string, data: { frontal: string | null; profile: string | null; body: string | null }): Promise<void> => {
        await apiClient.put(`/clone/${cloneId}/visual`, data);
    },

    // Update rights/privacy (Step 5)
    updateRights: async (cloneId: string, data: { is_public: boolean; allow_training: boolean }): Promise<void> => {
        await apiClient.put(`/clone/${cloneId}/rights`, data);
    },

    // Activate clone (Step 6) - starts async processing
    activate: async (cloneId: string): Promise<{ state: string; message: string }> => {
        const response = await apiClient.post(`/clone/${cloneId}/activate`);
        return response.data;
    },
};

export default cloneApi;


